docs/developer/decision-records/2025-02-05-state-machine-retry-processor-refactor/README.md

```diff
@@ -0,0 +1,159 @@
+                transitionToTerminated(t);
+                if (t.getType() == PROVIDER) {
+                    transitionToDeprovisioning(t);
+                }
+            })
+            .onFailure((t, throwable) -> transitionToTerminating(t, throwable.getMessage(), throwable))
+            .onFatalError((n, failure) -> transitionToTerminated(n, failure.getFailureDetail()))
+            .onRetryExhausted(this::transitionToTerminated)
+            .execute("send transfer termination to " + process.getCounterPartyAddress());
+}
+```
+
+In this case the `sendTransferTerminationMessage` is instantiating a new `RetryProcess` (in the `dispatch` method) with
+a different set of handlers (that are pretty much duplicated from the ones on top).
+
+This design doesn't permit the code to be expressed in a clear and fluent representation, in addition to requiring duplicated
+error handler.
+
+The current design has also additional flaws, as:
+- `onFatalError` and `onRetryExhausted` handled in a different way, but in fact they should be the same, because a "retry
+  exhausted" error can be considered a fatal error, an event that must stop retry
+- a lot of logic is duplicated in the `RetryProcess` subclasses
+- hard to understand abstract design
+
+The current implementation will be deprecated, and a new implementation will be provided to be used in parallel at first,
+until the deprecation expires and the old one can be deleted.
+
+## Approach
+
+### Concepts
+
+- `Process`: a function that given an `Entity` and an input can execute a sync or async operation and returns the `Entity`
+  and the output of the operation into an enclosing type
+    - the enclosing type is `CompletableFuture`, this because it permits to handle sync and async operation interchangeably.
+- `RetryProcessor` the component that will permit to set a chain of `Process`es that can be executed. Based on the result
+  of the operation and the retry conditions it will call one of the handlers:
+    - `onSuccess`: is called when all the processors execute successfully
+    - `onFailure`: is called when there's a failure but the operation is meant to be retries, depending on how many retries
+      the entity already went through and what is the configured `retryLimit`
+    - `onFinalFailure`: is called when retries are expired or an unrecoverable error happens
+      the processor returns true` if the at least one process gets executed or `false` if not (such case happens when a retry
+      strategy is waiting to be satisfied and processing will happen again on the next state machine iteration)
+
+Please note that all the `Process`es need to be idempotent, this will permit intermediate failures to be
+
+### Design
+
+#### ProcessContext
+It is a simple `record` that wraps an entity instance and a content that will be used as input and output of the `Process`:
+```java
+public record ProcessContext<E extends StatefulEntity<E>, C>(E entity, C content) { }
```

Comment:
nit: maybe an example would help understand the concept better, especially that of the `content`

Reply:
the example is available in the next section, I added a statement pointing to that.